### ISSUE TITLE
Fish shell freezing, spawning new processes at 100% CPU even when terminal closed

### ISSUE THREAD

**Issue by madsem** [netlify/cli#2934]
**Describe the bug**

I have signed up for Netlify recently and a couple of days ago I've installed netlify-cli.
Shortly after I noticed my terminal becoming slower and slower, until yesterday it was so bad that I couldn't use it anymore.

The terminal window was freezing, extreme lag when entering commands.
Today I used the Mac activity monitor, and saw that there were random fish processes spawned every couple of seconds,
those processes took between 50% -120% of CPU.

Even after a system restart, and closing my Terminal completely, these processes kept spawning...

I've started to debug this by process of elimination, because I wasn't sure what it was.
Luckily there were only two things I've installed recently, which was Starship and Netlify CLI.

Only after I disabled this line in my `config.fish`:
```fish
# The next line updates PATH for Netlify's Git Credential Helper.
test -f '/Users/madsem/Library/Preferences/netlify/helper/path.fish.inc' && source '/Users/madsem/Library/Preferences/netlify/helper/path.fish.inc'
```
everything went back to normal again after I restarted my terminal (ITerm2)

Voila, no more runaway processes, my machine runs silent and I can hear my own thoughts again.

**Configuration**

❯ npx envinfo --system --binaries --npmPackages netlify-cli --npmGlobalPackages netlify-cli
Need to install the following packages:
  envinfo
Ok to proceed? (y) y

  System:
    OS: macOS 11.4
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Memory: 8.94 GB / 32.00 GB
    Shell: 3.2.2 - /usr/local/bin/fish
  Binaries:
    Node: 16.0.0 - /usr/local/bin/node
    npm: 7.19.1 - /usr/local/bin/npm
  npmGlobalPackages:
    netlify-cli: 4.1.13

fish, version 3.2.2

**Expected behavior**

That there are no random processes spawned that shut down my entire machine :)

**CLI Output**

N/A

**Comment by overlordofmu** [comment]
@madsem, would you be willing to share some more details about the processes being spawned? Also, what version of the Fish shell is being used?

The logical and operator using the double ampersand (`&&`) wasn't supported until version 3.0.0 of Fish is it seems and I'm wondering if the test is command is being backgrounded because the shell version doesn't like the double ampersand:

https://superuser.com/questions/446930/chain-fish-commands-via-or

**Comment by madsem** [comment]
Hey @overlordofmu 

sure!

Version is already part of the original post above, but it is `3.2.2`. So pretty much the latest.
I never have problems using the double ampersand when manually entering commands.

There's really not much more info I could share, when opening the Activity Monitor on mac, you saw 1-2 processes called `fish` being spawned with new PIDs every few seconds. The processes most of the time, had 100%+ CPU usage and after a full day it was so bad that the entire Terminal froze.

Even restart of computer, restart of terminal did not help. The only thing that did was removing the above line entirely.

I tried your suggestion, as soon as I entered the line which sources the `.inc` my terminal starts lagging.
You can already see, entering any command has a delay of ~1sec and my macbook's fan starts spinning loudly.

Here is my entire `config.fish`:
```fish
# disable greeting
set fish_greeting

# if there's a .secrets file, source that
set secrets ~/.secrets
if test -e secrets
    source secrets
end

# source global envs / exports
source ~/.config/fish/envs.fish

# source aliases
if status --is-login
    source ~/.config/fish/aliases.fish
end

# source abbreviations
if status --is-login
    source ~/.config/fish/abbreviations.fish
end

# add entries from paths file to $PATH
source ~/.config/fish/paths.fish
for entry in $entries
    set -gx PATH $entry $PATH
end

# activate Docker shell completion
set -l dockerCompletions /Applications/Docker.app/Contents/Resources/etc/docker.fish-completion
if status --is-login
    and test -e $dockerCompletions
    source $dockerCompletions
end


# Fisher Package Manager
# https://github.com/jorgebucaran/fisher
if not functions -q fisher
    set -q XDG_CONFIG_HOME
    or set XDG_CONFIG_HOME ~/.config
    curl https://git.io/fisher --create-dirs -sLo $XDG_CONFIG_HOME/fish/functions/fisher.fish
    fish -c fisher
end


# shell integration
source ~/.iterm2_shell_integration.fish

# tabtab source for packages
# uninstall by removing these lines
[ -f ~/.config/tabtab/__tabtab.fish ]; and . ~/.config/tabtab/__tabtab.fish; or true

# Starship prompt init
starship init fish | source

# Adding this causes instant key lag, if left for a while the fan starts spinning
source '/Users/madsem/Library/Preferences/netlify/helper/path.fish.inc'
```


PS:
I just upgraded my fish shell to the latest release `fish, version 3.3.1`, same results.

**Comment by madsem** [comment]
Here are two screen recordings.

Without the config line in question:

https://user-images.githubusercontent.com/3166991/125921850-d410934b-989c-4d64-b982-9e67b3971f49.mov


And here after I added it to my config, and restarted my terminal:

https://user-images.githubusercontent.com/3166991/125921951-8e45f3ed-cfee-4837-b617-4cf2dbb32472.mov

**Comment by madsem** [comment]
What I can't really show is how these processes affect my terminal. But it really lags so badly that after a while you type, wait 2-5 seconds until your command appears letter by letter.

**Comment by madsem** [comment]
PSPS: Feel free to check out my dotfiles, with the entire fish config:
https://github.com/madsem/dotfiles/tree/master/.config/fish

**Comment by JGAntunes** [comment]
Hey @madsem 👋 

Thanks for the thorough report. If I could bother you for just a couple more things to help us debug this.

Could you share your `$PATH` output (i.e. `echo $PATH`) after sourcing the script?

Also, mind sharing the contents of the helper directory - `ls -la /Users/madsem/Library/Preferences/netlify/helper` - and the content of the `path.fish.inc` file - `cat /Users/madsem/Library/Preferences/netlify/helper/path.fish.inc`?

**Comment by madsem** [comment]
Hey @JGAntunes 

Here's my $PATH without:
```
❯ echo $PATH
/Users/madsem/.composer/vendor/bin /usr/local/sbin/ /Users/madsem/Library/Preferences/netlify/helper/bin /usr/local/bin /usr/bin /bin /usr/sbin /sbin /Library/Apple/usr/bin
```

Here after I added the line in question
```
❯ echo $PATH
/Users/madsem/Library/Preferences/netlify/helper/bin /Users/madsem/.composer/vendor/bin /usr/local/sbin/ /usr/local/bin /usr/bin /bin /usr/sbin /sbin /Library/Apple/usr/bin
```

```
❯ ls -la /Users/madsem/Library/Preferences/netlify/helper
total 16
drwxr-xr-x  5 madsem  staff  160 Jul  8 16:14 .
drwx------  4 madsem  staff  128 Jul 10 10:34 ..
drwxr-xr-x  3 madsem  staff   96 Jul  8 16:14 bin
-rw-r--r--  1 madsem  staff  138 Jul  8 16:14 git-config
-rw-r--r--  1 madsem  staff  103 Jul  8 16:14 path.fish.inc
```

```
❯ cat /Users/madsem/Library/Preferences/netlify/helper/path.fish.inc
#!/usr/bin/env fish

set -U fish_user_paths (dirname (status --current-filename))/bin $fish_user_paths
```

And again, as soon as I added 
```fish
# Adding this causes instant key lag, if left for a while the fan starts spinning
source '/Users/madsem/Library/Preferences/netlify/helper/path.fish.inc'
```
to the end of my fish.config, the input lag starts...

**Comment by JGAntunes** [comment]
Great, thanks for the details 👍 I'll try to reproduce this on my end and report back.

**Comment by JGAntunes** [comment]
Hey @madsem 👋 apologies for the delay looking into this. #3017 holds a fix for what I believe is the problem here.

Our utility script is constantly appending the lm binary to the user `$fish_user_paths`. For example, on my end I got the following:

```sh
> echo $fish_user_paths
/Users/jgantunes/.fzf/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin /Users/jgantunes/Library/Preferences/netlify/helper/bin
```

Depending on your particular fish config this clutter might be more or less "impactful" if you happen to have scripts or utilities that depend on this variable (hence why I believe your CPU ends up sky rocketing).

Mind printing your current `$fish_user_paths`?  Depending on its state, even after the fix, we might need to clean it up 👍

**Comment by JGAntunes** [comment]
The fix landed in v5.2.9 via - https://github.com/netlify/cli/pull/3018. Could you try it and report back? You might need to force the helper re-install via `netlify lm:install --force`.

Meanwhile if your `$fish_user_paths` is cluttered I've put together a quick function that should be able to clear it out for you:

```fish
function remove_from_path
  set -l path $argv
  set -l new_user_paths
  for i in $fish_user_paths
    if not test "$i" = "$path"
      set new_user_paths $i $new_user_paths
    end
  end
  set -U fish_user_paths $new_user_paths
end
```
You can set it up for your current session only or put it under `.config/fish/functions` if you think its useful.

```fish
$ remove_from_path /Users/(whoami)/Library/Preferences/netlify/helper/bin
```

Let us know if that did the trick.

**Comment by erezrokah** [comment]
Closing this per https://github.com/netlify/cli/issues/2934#issuecomment-887730974
Please comment if still an issue

**Comment by madsem** [comment]
Hey @JGAntunes sorry for late reply, I have been swamped over here.

Thanks a lot for the function, that was extremely helpful! :)
The user paths were indeed completely cluttered, so many entries in fact that my terminal cut off the output after running `echo $fish_user_paths`.

I have uninstalled Large Media in the meantime and we went with Cloudinary's Netlify CMS integration. But I'm sure your fix will benefit a lot of users here :)
